libia2/include/ia2.h

```diff
@@ -226,28 +228,125 @@ static int insecure_pkey_mprotect(void *ptr, size_t len, int prot, int pkey) {
+                     ",%%rdi\n"                                                \
+                     "push %%rbp\n"                                            \
+                     "# allocate a stack; puts its top in rax\n"               \
+                     "call allocate_stack\n"                                   \
+                     "mov %%rax, %%rbp"                                        \
+                     "# zero eax/ebx/ecx/edx\n"                                \
+                     "xor %%eax,%%eax\n"                                       \
+                     "mov %%eax,%%ebx\n"                                       \
+                     "mov %%eax,%%ecx\n"                                       \
+                     "mov %%eax,%%edx\n"                                       \
+                     "# read old pkru\n" IA2_RDPKRU                            \
+                     "\n"                                                      \
+                     "# save pkru in r12d\n"                                   \
+                     "mov %%eax,%%r12d\n"                                      \
+                     "# write new pkru"                                        \
+                     "mov_pkru_eax " #i "\n" IA2_WRPKRU                        \
+                     "\n"                                                      \
+                     "mov ia2_stackptr_" #i                                    \
+                     "@GOTTPOFF(%%rip), %%r11\n"                               \
+                     "cmpq $0x0,%%fs:(%%r11)\n"                                \
+                     "je .Lfresh_init" #i                                      \
+                     "\n"                                                      \
+                     "mov $" #i                                                \
+                     ", %%rdi\n"                                               \
+                     "call ia2_reinit_stack_err\n"                             \
+                     ".Lfresh_init" #i                                         \
+                     ":\n"                                                     \
+                     "# store the stack addr in the stack pointer\n"           \
+                     "mov %%rbp, %%fs:(%%r11)\n"                               \
+                     "# restore old pkru\n"                                    \
+                     "mov %%r12d,%%eax\n" IA2_WRPKRU                           \
+                     "\n"                                                      \
+                     "pop %%rbp\n"                                             \
+                     :                                                         \
+                     :                                                         \
+                     : "rdi", "rax", "rbx", "rcx", "rdx", "r11", "r12");       \
```

Comment:
Can you just clobber rbp and not add the pushes?

Reply:
Also do we have to put all the caller-saved regs in the clobber list because we make function calls? I don't remember extended asm well enough off the top of my head...

Reply:
Clobbering rbp is somewhat contentious: see [some comments here](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104805). We could use another register instead, e.g. r10.

I believe you're right, we should mark caller-save regs as clobbered. That's kind of a pain, but we can also relatively easily just pass a single variable that we ask gcc to assign a register into our asm and move the call to `allocate_stack` out. I'll do that.

Reply:
hmmm yeah I'd be in favor of both pulling out the call and using a different reg

Reply:
Done.